Redis Appliance

benbjohnson opened the pull request:
## Overview

This pull request adds a `redis` resource that can be added to applications. It's currently only a single process and does not provide clustering or replication.


### Usage

The resource can be added via the `flynn` CLI:

```sh
$ flynn resource add redis
```

This creates a new app for a redis process and sets the `FLYNN_REDIS` and `FLYNN_REDIS_URL` environment variables. The user can also connect to the redis app via the CLI:

```sh
$ flynn redis redis-cli
```

josephglanville commented:
This looks great so far.

Are you planning to implementing replication failover using the Redis Sentinel thing or something based on `discoverd` instead similar to the existing PostgreSQL state machine?

titanous commented:
@josephglanville This will be merged without replication as it's useful to the many people who use Redis as some form of cache server (and it's useful for development as well). We'll add replication in the future, but I don't think we can re-use the PostgresSQL state machine because last I checked, Redis does not have a synchronous replication mode.

josephglanville commented:
Yes, async only which means it can't be made fully consistent. Also makes sense to merge without replication - was just wondering what the approach is going to be.

benbjohnson commented:
I was having git rebase issues but I eventually fixed it by cherry-picking and rebasing every commit individually. @titanous should i go ahead and merge this?

benbjohnson commented:
nm. There's a failing unit test case because of a missing `redis-server` binary. I'll see if I can switch it to use the docker image instead of the local binary.

titanous commented:
I'm going to do a full review of this.

titanous commented on appliance/redis/process.go:
```diff
@@ -0,0 +1,436 @@
+	}
+}
+
+// pingWait continually pings a server until
```
Incomplete comment.

benbjohnson replied:
Fixed in 478e3a3.

josephglanville commented on bootstrap/manifest_template.json:
```diff
@@ -125,12 +125,72 @@
+      "name": "redis",
+      "meta": {"flynn-system-app": "true"}
+    },
+    "action": "run-app",
```
You should be able to use `deploy-app` and deploy Redis after `controller` is up and running. This will avoid the 2 stage process of `run-app` followed by `deploy-app`.

benbjohnson replied:
I cleaned up the manifest in 478e3a3. The `redis` steps weren't needed since they're done by `redis-api`.

josephglanville commented on bootstrap/manifest_template.json:
```diff
@@ -125,12 +125,72 @@
+          "env": {
+            "SINGLETON": "{{ .Singleton }}"
+          },
+          "resurrect": true
```
I don't think Redis needs to be ressurected. The once the `controller` has been successfully resurrected it will sync the Redis formations with the state of the world and fix everything up.

benbjohnson replied:
Removed from `redis-api` in 6a96cef.

josephglanville commented on bootstrap/run_app_action.go:
```diff
@@ -81,6 +81,12 @@ func (a *RunAppAction) Run(s *State) error {
 	interpolateRelease(s, a.Release)
 
 	for _, p := range a.Resources {
+		// Register provider without allocating resource.
```
This change doesn't seem to be used? All of the resources specified in the manifest have `URL` set.

benbjohnson replied:
I used it in earlier code before `add-provider`. Removed in 6a96cef.

josephglanville commented on bootstrap/manifest_template.json:
```diff
@@ -204,7 +269,9 @@
-    "resources": [{"name":"postgres", "url":"http://postgres-api.discoverd/databases"}]
+    "resources": [
+      {"name":"postgres", "url":"http://postgres-api.discoverd/databases"}
+    ]
```
Unintentional whitespace change?

benbjohnson replied:
Fixed in 6a96cef.

benbjohnson commented:
@titanous I added docs, cleaned up `redis` items in the manifest, added support for passwords in the redis appliance, and cleaned up some environment variables returned by the provider. I still need to fix the unit tests.

josephglanville commented on bootstrap/manifest_template.json:
```diff
@@ -131,6 +131,34 @@
+      "web": 2
+    }
+  },
+  {
```
If you don't follow `run-app` with `add-app` the application won't be added to the `controller`. You should probably use `deploy-app` instead though and move this until after the controller is fully setup as it's not required to get the controller going.

benbjohnson replied:
I tried moving `redis-api` and `redis-api-wait` down below the controller and I'm getting strange results. Here's the manifest file:

https://gist.github.com/benbjohnson/3b22c55f0bf1b30e119d

I'm seeing `deploy-app redis-api` during bootstrap but then it hangs on `wait redis-api-wait`. I checked the logs and I don't see any `redis-api` process running.

https://gist.github.com/anonymous/37681124ae54ac1738c8

josephglanville replied:
Yep, because the scheduler isn't fired up yet. Move down to where `blobstore` and friends are. The controller needs to be fully setup before `deploy-app` works.

I can probably test this today if my internet is fixed but it should work.

titanous commented on cli/redis.go:
```diff
@@ -0,0 +1,76 @@
+usage: flynn redis redis-cli [--] [<argument>...]
+
+Options:
+	-q, --quiet        don't print progress
```
Is this option used?

benbjohnson replied:
No, removed in c6b8b6a.

titanous commented on cli/redis.go:
```diff
@@ -0,0 +1,76 @@
+	switch {
+	case args.Bool["redis-cli"]:
+		return runRedisCLI(args, client, config)
+	}
```
Default case should probably error.

titanous replied:
Actually, maybe docopt handles that, worth testing.

benbjohnson replied:
Tested, `docopt` handles it.

titanous commented on docs/content/redis.html.md:
```diff
@@ -0,0 +1,38 @@
+# Redis
+
+The Flynn Redis appliance provides Redis 2.8 in a single process configuration.
+
```
Should note that data is ephemeral, this is intended for caching and development usage.

benbjohnson replied:
Clarified in c6b8b6a.

titanous commented on docs/content/redis.html.md:
```diff
@@ -0,0 +1,38 @@
+### Connecting to a console
+
+To connect to a console for the database, run `flynn redis redis-cli`. This does
+not require the Flynn client to be installed locally or firewall/security
```
s/Flynn/Redis/

benbjohnson replied:
Fixed in c6b8b6a.

titanous commented on controller/provider.go:
```diff
@@ -18,6 +18,7 @@ func NewProviderRepo(db *postgres.DB) *ProviderRepo {
 
 func (r *ProviderRepo) Add(data interface{}) error {
 	p := data.(*ct.Provider)
+
```
Extraneous whitespace change?

benbjohnson replied:
Removed in c6b8b6a.

titanous commented:
`flynn export/import` should end up creating a new redis instance for the app and override the env vars (similar to postgres without any data movement).

titanous commented on appliance/redis/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+
+RUN apt-get update &&\
+    apt-get dist-upgrade -y &&\
+    apt-get -y install curl sudo &&\
```
Is this line necessary?

benbjohnson replied:
Removed in c6b8b6a. Updated the `Dockerfile` to not use `sudo`.

titanous commented on appliance/redis/cmd/flynn-redis-api/main.go:
```diff
@@ -0,0 +1,276 @@
+		return
+	}
+
+	h.Logger.Info("creating release:", "artifactID", artifact.ID)
```
s/://

titanous replied:
s/artifactID/artifact.id/

benbjohnson replied:
Fixed in c6b8b6a.

titanous commented on appliance/redis/cmd/flynn-redis/main.go:
```diff
@@ -0,0 +1,218 @@
+	m.Process.Password = os.Getenv("REDIS_PASSWORD")
+
+	if m.Process.Password != "" {
+		m.Logger.Info("set password", "password", m.Process.Password)
```
The password should not be logged.

benbjohnson replied:
Removed in c6b8b6a.

titanous commented on appliance/redis/handler.go:
```diff
@@ -0,0 +1,61 @@
+	info, err := h.Process.Info()
+	if err != nil {
+		// Log the error, but don't return a 500. We will always have some
+		// information to return, but postgres may not be online.
```
s/postgres/redis/

benbjohnson replied:
Fixed in c6b8b6a.

titanous commented on appliance/redis/process.go:
```diff
@@ -0,0 +1,439 @@
+	p.stopping.Store(true)
+
+	// Attempt to kill via escalating signals.
+	for _, sig := range []os.Signal{syscall.SIGINT, syscall.SIGQUIT, syscall.SIGKILL} {
```
Does redis respond differently to these signals? Postgres does.

benbjohnson replied:
These were the incorrect signals. Changed to `SIGTERM` (graceful shutdown) and `SIGSEV` (hard shutdown) in c6b8b6a.

http://redis.io/topics/signals

benbjohnson commented:
@titanous I added support for flynn import/export in c6b8b6a.
